[SPARK-3162] [MLlib] Add local tree training for decision tree regressors

smurching opened the pull request:
## What changes were proposed in this pull request?

### Overview
This PR adds local tree training for decision tree regressors as a first step for addressing [SPARK-3162](https://issues.apache.org/jira/browse/SPARK-3162): train decision trees locally when possible.

See [this design doc](https://docs.google.com/document/d/1baU5KeorrmLpC4EZoqLuG-E8sUJqmdELLbr8o6wdbVM/edit) (in particular the [local tree training section](https://docs.google.com/document/d/1baU5KeorrmLpC4EZoqLuG-E8sUJqmdELLbr8o6wdbVM/edit#heading=h.eiteb09wxsjf)) for detailed discussion of the proposed changes.

Distributed training logic has been refactored but only minimally modified; the local tree training implementation leverages existing distributed training logic for computing impurities and splits. This shared logic has been refactored into `...Utils` objects (e.g. `SplitUtils.scala`, `ImpurityUtils.scala`). 

### How to Review

Each commit in this PR adds non-overlapping functionality, so the PR can be reviewed commit-by-commit.

Changes introduced by each commit:
1. Adds new data structures for local tree training (`FeatureVector`, `TrainingInfo`)
2. Adds shared utility methods for computing splits/impurities (`SplitUtils`, `ImpurityUtils`, `AggUpdateUtils`), largely copied from existing distributed training code in `RandomForest.scala`.
3. Unit tests for split/impurity utility methods (`TreeSplitUtilsSuite`)
4. Updates distributed training code in `RandomForest.scala` to depend on the utility methods introduced in 2.
5. Adds local tree training logic (`LocalDecisionTree`) 
6. Local tree unit/integration tests (`LocalTreeUnitSuite`, `LocalTreeIntegrationSuite`)

## How was this patch tested?
No existing tests were modified. The following new tests were added (also described above):
* Unit tests for new data structures specific to local tree training (`LocalTreeDataSuite`, `LocalTreeUtilsSuite`)
* Unit tests for impurity/split utility methods (`TreeSplitUtilsSuite`)
* Unit tests for local tree training logic (`LocalTreeUnitSuite`)
* Integration tests verifying that local & distributed tree training produce the same trees (`LocalTreeIntegrationSuite`)

smurching committed:
Add data structures for local tree training & associated tests (in LocalTreeDataSuite):
    * TrainingInfo: primary local tree training data structure, contains all information required to describe state of
    algorithm at any point during learning
    * FeatureVector: Stores data for an individual feature as an Array[Int]

smurching committed:
Add utility methods used for impurity and split calculations during both local & distributed training:
 * AggUpdateUtils: Helper methods for updating sufficient stats for a given node
 * ImpurityUtils: Helper methods for impurity-related calcluations during node split decisions
 * SplitUtils: Helper methods for choosing splits given sufficient stats

NOTE: Both ImpurityUtils and SplitUtils primarily contain code taken from RandomForest.scala, with slight modifications.
Tests for SplitUtils are contained in the next commit.

smurching committed:
Add test suites for utility methods used during best-split computation:
 * TreeSplitUtilsSuite: Test suite for SplitUtils
 * TreeTests: Add utility method (getMetadata) for TreeSplitUtilsSuite

 Also add methods used by these tests in LocalDecisionTree.scala, RandomForest.scala

smurching committed:
Update RandomForest.scala to use new utility methods for impurity/split calculations

smurching committed:
Add local decision tree training logic

smurching committed:
Add local decision tree unit/integration tests

AmplabJenkins commented:
Can one of the admins verify this patch?

smurching commented:
@WeichenXu123 would you be able to take an initial look at this?

smurching commented on mllib/src/main/scala/org/apache/spark/ml/tree/impl/LocalDecisionTreeUtils.scala:
```diff
@@ -0,0 +1,59 @@
+    val numFeatures = rowStore(0).length
+    require(numFeatures > 0, "Local decision tree training requires numFeatures > 0.")
+    // Return the transpose of the rowStore matrix
+    0.until(numFeatures).map { colIdx =>
```
TODO: replace this with `rowStore.transpose`, which is more memory efficient (iterates over each row once, allowing for rows of the original matrix to be GC'd during the transpose operation).

WeichenXu123 commented:
@smurching Does it still WIP ? If done remove "[WIP]", I will begin review, thanks!

smurching commented:
Thanks! I'll remove the WIP. To clear things up for the future, I'd thought [WIP] was the appropriate tag for a PR that's ready for review but not ready to be merged (based on https://spark.apache.org/contributing.html) -- have we stopped using the WIP tag?

jkbradley commented:
add to whitelist

SparkQA commented:
**[Test build #82557 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82557/testReport)** for PR 19433 at commit [`9a7174e`](https://github.com/apache/spark/commit/9a7174ed4a62033abfe2325dc1a8c5850e07f5f3).

SparkQA commented:
**[Test build #82557 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82557/testReport)** for PR 19433 at commit [`9a7174e`](https://github.com/apache/spark/commit/9a7174ed4a62033abfe2325dc1a8c5850e07f5f3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class LocalTreeIntegrationSuite extends SparkFunSuite with MLlibTestSparkContext `
  * `class LocalTreeUtilsSuite extends SparkFunSuite `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82557/
Test FAILed.

smurching commented:
The failing tests (in `DecisionTreeSuite`) fail because we've historically handled

a) splits that have 0 gain 

differently from

b) splits that fail to achieve user-specified minimum gain (`metadata.minInfoGain`) or don't meet minimum instance-counts per node (`metadata.minInstancesPerNode`).

Previously we'd create a leaf node with valid impurity stats in case a) and invalid impurity stats in case b). This PR creates a leaf node with invalid impurity stats in both cases.

As a fix I'd suggest creating a `LeafNode` with correct impurity stats in case a), but with the `stats.valid` member set to `false` to indicate that the node should not be split.

smurching committed:
Fix tests, remove perf test in LocalTreeIntegrationSuite, use Scala transpose in LocalDecisionTreeUtils.
Changes made to fix tests:
 * Return correct impurity stats for splits that achieved a gain of 0 but didn't violate user-specified constraints on min info gain or min
 instances per node
 * Previously, ImpurityStats.impurity was set incorrectly in ImpurityStats.getInvalidImpurityStats(), requiring a correction in LearningNode.toNode.
   This commit fixes the issue by directly setting impurity = -1 in getInvalidSplits()

SparkQA commented:
**[Test build #82570 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82570/testReport)** for PR 19433 at commit [`abc86b2`](https://github.com/apache/spark/commit/abc86b2042e0fd42cc0e9fe20cf79967b16e9779).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82570/
Test FAILed.

smurching commented:
The failing SparkR test (which compares `RandomForest` predictions to hardcoded values) fails not due to a correctness issue but (AFAICT) because of an implementation change in best-split selection. 

In this PR we recompute parent node impurity stats when considering each split for a feature, instead of computing parent impurity stats once per feature (see this by comparing `RandomForest.calculateImpurityStats` in Spark master and `ImpurityUtils.calculateImpurityStats` in this PR).

The process of repeatedly computing parent impurity stats results in slightly different results at each iteration due to Double precision limitations. This in turn can cause different splits to be selected (e.g. if two splits have mathematically equal gains, Double precision limitations can cause one split to have a higher/smaller gain than the other, influencing tiebreaking).

smurching committed:
Update calculateImpurityStats  to avoid recomputing parent impurity stats during best split selection

SparkQA commented:
**[Test build #82652 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82652/testReport)** for PR 19433 at commit [`5c29d3d`](https://github.com/apache/spark/commit/5c29d3d1e899c8d311633c4d763b57e42a26c660).

SparkQA commented:
**[Test build #82652 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82652/testReport)** for PR 19433 at commit [`5c29d3d`](https://github.com/apache/spark/commit/5c29d3d1e899c8d311633c4d763b57e42a26c660).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82652/
Test FAILed.

smurching committed:
Merge branch 'master' into pr-splitup

WeichenXu123 commented:
I made a rough pass. I have only a few issues for now, I haven't go into code details:

- The `colStoreInit` currently ignore the `subsampleWeights`, it should be used, isn't it ? I read your doc, in the higher level, the local training will be used to train sub-trees as parts of the global distributed training, `subsampleWeights` should be important info. and here it will train only single tree so `subsampleWeights` only contains one element, does we still need use `BaggedPoint` structure ? 

- The logic of training for regression and for classification will be the same I think, only impurity difference but do not influence the code logic.

- The key idea is to use the columnar storage format for features, is the purpose to improve memory cost & cache locality when finding best splits ? I see the code will do some reordering operation on feature values and use indices, but I haven't go into details. It's a complex part I need more time to review.

- Maybe we can support multithreads in local training, what do you think about it ?

smurching commented:
Thanks for the comments!

- Yep, feature subsampling is necessary for using local tree training in distributed training. I was thinking of adding subsampling in a follow-up PR. You're right that we don't need to pass an array of `BaggedPoints` to local tree training; we should just pass an array of `subsampleWeights` (weights for the current tree) and an array of `TreePoints`. I'll push an update for this.

- Agreed that the logic for classification will be the same but with a different impurity metric. I can add support for classification & associated tests in a follow-up PR.

- IMO the primary advantage of the columnar storage format is that it'll eventually enable improvements to best split calculations; specifically, for continuous features we could sort the unbinned feature values and consider every possible threshold.  There are also the locality & memory advantages described in the design doc. In brief, `DTStatsAggregator` stores a flat array partitioned by (feature x bin). If we can iterate through all values for a single feature at once, most updates to `DTStatsAggregator`will occur within the same subarray.

- Multithreading could be a nice way to increase parallelism since we don't use Spark during local tree training. I think we could add it in a follow-up PR.

smurching commented:
Sorry, realized I conflated feature subsampling and `subsampleWeights` (instance weights for training examples). IMO feature subsampling can be added in a follow-up PR, but `subsampleWeights` should go in this PR.

smurching committed:
Fix test bug where instanceWeights weren't properly passed to update methods

SparkQA commented:
**[Test build #82717 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82717/testReport)** for PR 19433 at commit [`c9a8e01`](https://github.com/apache/spark/commit/c9a8e01cead78d2ea6eb6a9ffa007314cbbcf60a).

SparkQA commented:
**[Test build #82717 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82717/testReport)** for PR 19433 at commit [`c9a8e01`](https://github.com/apache/spark/commit/c9a8e01cead78d2ea6eb6a9ffa007314cbbcf60a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82717/
Test PASSed.

smurching committed:
Use per-training-example instance weights in local tree training

SparkQA commented:
**[Test build #82721 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82721/testReport)** for PR 19433 at commit [`93e17fc`](https://github.com/apache/spark/commit/93e17fc74958d4fa8f3bea38731ecec662e4ca66).

SparkQA commented:
**[Test build #82721 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82721/testReport)** for PR 19433 at commit [`93e17fc`](https://github.com/apache/spark/commit/93e17fc74958d4fa8f3bea38731ecec662e4ca66).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82721/
Test PASSed.

WeichenXu123 commented on mllib/src/main/scala/org/apache/spark/ml/tree/impl/AggUpdateUtils.scala:
```diff
@@ -0,0 +1,85 @@
+   */
+  private[impl] def updateParentImpurity(
+      statsAggregator: DTStatsAggregator,
+      col: FeatureVector,
```
Actually, `updateParentImpurity` has no relation with any feature column, but here you pass in the `feature` column only want to use the `indices` array, passing anyone feature column will be OK. But, this looks weird, maybe it can be better designed.

WeichenXu123 commented on mllib/src/main/scala/org/apache/spark/ml/tree/impl/AggUpdateUtils.scala:
```diff
@@ -0,0 +1,85 @@
+      label: Double,
+      featureIndex: Int,
+      featureIndexIdx: Int,
+      splits: Array[Array[Split]],
```
You only need to pass in the `featureSplit: Array[Split]`, don't pass all splits for all features.

smurching replied:
Good call, I'll make this change.

WeichenXu123 commented on mllib/src/main/scala/org/apache/spark/ml/tree/impl/TrainingInfo.scala:
```diff
@@ -0,0 +1,144 @@
+      from: Int,
+      to: Int,
+      split: Split,
+      allSplits: Array[Array[Split]]): BitSet = {
```
Ditto, you only need to pass in the featureSplit: Array[Split], don't pass all splits for all features.

WeichenXu123 reviewed (commented):
I made a deeper pass review, Later I will put more thoughts on the columnar feature storage design. Thanks!

WeichenXu123 commented:
@smurching I found some issues and have some thoughts on the columnar features format:

- In your doc, you said "Specifically, we only need to store sufficient stats for each bin of a single feature, as opposed to each bin of every feature", BUT, current implementation, you still allocate space for all features when computing: -- see `DTStatsAggregator` implementation, you pass `featureSubset = None` so `DTStatsAggregator` will allocate space for every features. According to your purpose, you should pass `featureSubset = Some(Array(currentFeatureIndex))`.

- Current implementation still use binnedFeatures. You said in future it will be improved to sort feature values for continuous feature (for more precise tree training), if you want to consider every possible thresholds, you need hold rawFeatures instead of binnedFeatures in the columnar feature array, and in each split range offset, you need sort every continuous features. Is this the thing you want to do in the future ? This will increase calculation amount.

- For current implementation(using binnedFeature) , there is no need to sort continuous features inside each split offset. So the `indices` for each feature is exactly the same. In order to save memory, I think these `indices` should be shared, no need to create separate indices array for each features. Even if you add the improvements for continuous features mentioned above, you can create separate `indices` array for **only** continuous features, the categorical features can still share the same `indices` array.

- About locality advantage of columnar format, I have some doubts. Current implementation, you do not reorder the `label` and `weight` array, access `label` and `weight` value need use `indices`, when calculating `DTStat`, this break locality. (But I'm not sure how much impact to perf this will bring).

- About the overhead of columnar format: when making reordering (when get new split, we need reorder left sub-tree samples into front), so you need reordering on each column, and at the same time, update the `indices` array. But, if we use row format, like:
`Array[(features, label, weight)]`, reordering will be much easier, and do not need indices.
So, I am considering, whether we can use row format, but at the time when we need `DTStatsAggregator` computation, copy the data we need from the row format into columnar format array (only need to copy rows between sub-node offset and only copy the sampled features if using feature subsampling).

smurching commented on mllib/src/main/scala/org/apache/spark/ml/tree/impl/FeatureVector.scala:
```diff
@@ -0,0 +1,148 @@
+    // gives us the split bit value for each instance based on the instance's index.
+    // We copy our feature values into @tempVals and @tempIndices either:
+    // 1) in the [from, numLeftRows) range if the bit is false, or
+    // 2) in the [numBitsNotSet, to) range if the bit is true.
```
Will change this, thanks for the catch!

smurching commented on mllib/src/main/scala/org/apache/spark/ml/tree/impl/LocalDecisionTree.scala:
```diff
@@ -0,0 +1,250 @@
+      val activeNodes: Array[LearningNode] =
+        computeBestSplits(trainingInfo, labels, metadata, splits)
+      // Filter active node periphery by impurity.
+      val estimatedRemainingActive = activeNodes.count(_.stats.impurity > 0.0)
```
Agreed on using `isLeaf` instead of checking for positive impurity, thanks for the suggestion.

AFAICT at this point in the code `activeNodes` actually does refer to the nodes in the current level; the children of nodes in `activeNodes` are the nodes in the next level, and are returned by `computeBestSplits`. I forgot to include the return type of `computeBestSplit` in its method signature, which probably made this more confusing - my mistake.

WeichenXu123 replied:
Yes. Sorry for confusing you. The change what I said was changing to:
```
val nextLevelNodes: Array[LearningNode] =
        computeBestSplits(trainingInfo, labels, metadata, splits)
```
Does it look more reasonable ?
And change the member name in `trainingInfo`:
`TrainingInfo.activeNodes` ==> `TrainingInfo.currentLevelNodes`

smurching replied:
Gotcha agreed on the naming change, how about `currentLevelActiveNodes`? Since only the non-leaf nodes from the current level are included.

WeichenXu123 replied:
Wait... I check the code here: `trainingInfo = trainingInfo.update(splits, activeNodes)` So it seems you do not filter out the leaf node from the "activeNodes"(which is actually the `nextLevelNode` I mentioned above).
So I think `TrainingInfo.activeNodes` is still possible to contains leaf node.

smurching replied:
Oh true -- I'll doc for `currentLevelActiveNodes` to say:
```
 * @param currentLevelActiveNodes  Nodes which are active (could still be split).
 *                                 Inactive nodes are known to be leaves in the final tree.
```

smurching commented on mllib/src/main/scala/org/apache/spark/ml/tree/impl/TrainingInfo.scala:
```diff
@@ -0,0 +1,144 @@
+ */
+private[impl] case class TrainingInfo(
+    columns: Array[FeatureVector],
+    instanceWeights: Array[Double],
```
Good call, I'll move `instanceWeights` outside `TrainingInfo`

smurching committed:
Respond to review comments:
* Move instanceWeights outside TrainingInfo
* Only pass a single array of splits (instead of an array of arrays of splits) when possible

SparkQA commented:
**[Test build #83025 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83025/testReport)** for PR 19433 at commit [`fd6cdbb`](https://github.com/apache/spark/commit/fd6cdbb66fd8beae944d0bd2029e8d62096747a8).

SparkQA commented:
**[Test build #83025 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83025/testReport)** for PR 19433 at commit [`fd6cdbb`](https://github.com/apache/spark/commit/fd6cdbb66fd8beae944d0bd2029e8d62096747a8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83025/
Test PASSed.

smurching commented:
@WeichenXu123 Thanks for the comments! I'll respond inline:

> In your doc, you said "Specifically, we only need to store sufficient stats for each bin of a single feature, as opposed to each bin of every feature", BUT, current implementation, you still allocate space for all features when computing: -- see DTStatsAggregator implementation, you pass featureSubset = None so DTStatsAggregator will allocate space for every features. According to your purpose, you should pass featureSubset = Some(Array(currentFeatureIndex)).

I like your proposed solution (pass `featureSubset = Some(Array(currentFeatureIndex))`). I'll go ahead & implement it.

> Current implementation still use binnedFeatures. You said in future it will be improved to sort feature values for continuous feature (for more precise tree training), if you want to consider every possible thresholds, you need hold rawFeatures instead of binnedFeatures in the columnar feature array, and in each split range offset, you need sort every continuous features. Is this the thing you want to do in the future ? This will increase calculation amount.

Yep, we'll have to pass raw (`Double`) continuous features to the local tree training methods, which will require them to accept an `Array[LabeledPoint]` instead of an `Array[TreePoint]` as input & increase memory usage (along with requiring us to store additional indices). 

We'll actually only have to run an `O(n log n)` sort on continuous feature values once (i.e. in the `FeatureVector` constructor), since once the continuous features are sorted we can update them as we would for categorical features when splitting nodes (in `O(n)` time) and they'll remain sorted.

> For current implementation(using binnedFeature) , there is no need to sort continuous features inside each split offset. So the indices for each feature is exactly the same. In order to save memory, I think these indices should be shared, no need to create separate indices array for each features. Even if you add the improvements for continuous features mentioned above, you can create separate indices array for only continuous features, the categorical features can still share the same indices array.

Agreed, I'll make this change.

> About locality advantage of columnar format, I have some doubts. Current implementation, you do not reorder the label and weight array, access label and weight value need use indices, when calculating DTStat, this break locality. (But I'm not sure how much impact to perf this will bring).

Yeah, I'm not sure if it'd be better to reorder the labels/weights arrays to achieve improved locality.
I think we could experiment with both, but I'd prefer to save that for a follow-up PR unless you or another reviewer think it'll make a big perf difference.

> About the overhead of columnar format: when making reordering (when get new split, we need reorder left sub-tree samples into front), so you need reordering on each column, and at the same time, update the indices array. But, if we use row format, like:
Array[(features, label, weight)], reordering will be much easier, and do not need indices.
So, I am considering, whether we can use row format, but at the time when we need DTStatsAggregator computation, copy the data we need from the row format into columnar format array (only need to copy rows between sub-node offset and only copy the sampled features if using feature subsampling).

This is an interesting idea, my main concern is that on the first iteration of local tree training we'd need to copy the entire training data matrix from row -> columnar format, which negates any memory savings we get from not using indices. I'm also concerned about the overhead of repeatedly copying data from row -> columnar format.

smurching committed:
WIP, renamed activeNodes -> currentLevelActiveNodes, WIP sharing memory by calling chooseSplit() directly in local tree impl

WeichenXu123 commented:
> We'll actually only have to run an O(n log n) sort on continuous feature values once (i.e. in the FeatureVector constructor), since once the continuous features are sorted we can update them as we would for categorical features when splitting nodes (in O(n) time) and they'll remain sorted.

Nice! so only one pass global sort, and then each split only need O(n) time copy.

smurching committed:
WIP

smurching committed:
WIP, about to replace dummyStatsAggregator with an ImpurityAggregator (instance that operates over an array of stats)

smurching committed:
More WIP, use raw ImpurityCalculator/ImpurityAggregator in LocalDecisionTree (saves an unnecessary array allocation)

smurching committed:
Revert "More WIP, use raw ImpurityCalculator/ImpurityAggregator in LocalDecisionTree (saves an unnecessary array allocation)"

This reverts commit 1e5db8a64c7423123d4c0d89b9611de0bc01438f.

smurching committed:
Rename dummyStatsAggregator -> currNodeStatsAgg

SparkQA commented:
**[Test build #83093 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83093/testReport)** for PR 19433 at commit [`ebade23`](https://github.com/apache/spark/commit/ebade2349153708c125fbeca900daec3d84c3513).

SparkQA commented:
**[Test build #83093 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83093/testReport)** for PR 19433 at commit [`ebade23`](https://github.com/apache/spark/commit/ebade2349153708c125fbeca900daec3d84c3513).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83093/
Test PASSed.

smurching committed:
Respond to review comments

smurching commented:
Made a few updates, here’s a quick summary/what I’d propose moving forward:

Right now:
* Shared row indices for all (categorical & continuous) features are stored & updated in `TrainingInfo`
* `LocalDecisionTree.computeBestSplits` computes best splits/sufficient stats for a single feature at a time
* A utility method (`LocalDecisionTreeUtils.updateArrayForSplit`) is used to sort both feature values and shared row indices

When we add support for raw continuous feature values:
* Add a subclass of `FeatureColumn` (e.g. `ContinuousFeatureColumn`) that stores and sorts its own array of row indices, pass these row indices to methods requiring them.

I also renamed `FeatureVector` to `FeatureColumn` since the former seemed like it’d confuse developers (`FeatureVector` sounds like a single data point)

SparkQA commented:
**[Test build #83104 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83104/testReport)** for PR 19433 at commit [`9cc6333`](https://github.com/apache/spark/commit/9cc63334900d9928f58e3bab5bd1be452d042a53).

SparkQA commented:
**[Test build #83104 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83104/testReport)** for PR 19433 at commit [`9cc6333`](https://github.com/apache/spark/commit/9cc63334900d9928f58e3bab5bd1be452d042a53).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83104/
Test PASSed.

WeichenXu123 commented on mllib/src/main/scala/org/apache/spark/ml/tree/impl/LocalDecisionTree.scala:
```diff
@@ -0,0 +1,255 @@
+    // For each node, select the best split across all features
+    trainingInfo match {
+      case TrainingInfo(columns: Array[FeatureVector],
+      nodeOffsets: Array[(Int, Int)], activeNodes: Array[LearningNode]) => {
```
`activeNodes` ==> `currentLevelNodes`

WeichenXu123 commented:
After discussion and modifications, I approve this PR overall. Ping @jkbradley Can you take a look now ?

smurching committed:
Fix indentation, remove unnecessary params from LocalDecisionTreeRegressor

SparkQA commented:
**[Test build #83219 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83219/testReport)** for PR 19433 at commit [`7efb1e0`](https://github.com/apache/spark/commit/7efb1e0b551235036d7e24dfe16f6fb9e517e503).

SparkQA commented:
**[Test build #83219 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83219/testReport)** for PR 19433 at commit [`7efb1e0`](https://github.com/apache/spark/commit/7efb1e0b551235036d7e24dfe16f6fb9e517e503).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83219/
Test PASSed.

smurching committed:
Remove redundant helper method in LocalTreeUnitSuite

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83310/
Test FAILed.

smurching commented:
jenkins retest this please

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83353/
Test FAILed.

JoshRosen commented:
jenkins retest this please

SparkQA commented:
**[Test build #83464 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83464/testReport)** for PR 19433 at commit [`3f72cc0`](https://github.com/apache/spark/commit/3f72cc0d92132f850c5eebe2686473bff111199f).

SparkQA commented:
**[Test build #83464 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83464/testReport)** for PR 19433 at commit [`3f72cc0`](https://github.com/apache/spark/commit/3f72cc0d92132f850c5eebe2686473bff111199f).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83464/
Test FAILed.

smurching commented:
jenkins retest this please

smurching committed:
Merge branch 'master' into pr-splitup

SparkQA commented:
**[Test build #83503 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83503/testReport)** for PR 19433 at commit [`3f72cc0`](https://github.com/apache/spark/commit/3f72cc0d92132f850c5eebe2686473bff111199f).

SparkQA commented:
**[Test build #83503 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83503/testReport)** for PR 19433 at commit [`3f72cc0`](https://github.com/apache/spark/commit/3f72cc0d92132f850c5eebe2686473bff111199f).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83503/
Test FAILed.

SparkQA commented:
**[Test build #83507 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83507/testReport)** for PR 19433 at commit [`b7e6e40`](https://github.com/apache/spark/commit/b7e6e40976612546b81d9775c194b274c146dc85).

SparkQA commented:
**[Test build #83507 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83507/testReport)** for PR 19433 at commit [`b7e6e40`](https://github.com/apache/spark/commit/b7e6e40976612546b81d9775c194b274c146dc85).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83507/
Test FAILed.

jkbradley commented:
CC @dbtsai  in case you're interested b/c of Sequoia forests

SparkQA commented:
**[Test build #3983 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3983/testReport)** for PR 19433 at commit [`b7e6e40`](https://github.com/apache/spark/commit/b7e6e40976612546b81d9775c194b274c146dc85).

SparkQA commented:
**[Test build #3983 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3983/testReport)** for PR 19433 at commit [`b7e6e40`](https://github.com/apache/spark/commit/b7e6e40976612546b81d9775c194b274c146dc85).
 * This patch **fails to generate documentation**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala:
```diff
@@ -852,6 +662,41 @@ private[spark] object RandomForest extends Logging {
+      node: LearningNode): (Split, ImpurityStats) = {
+    val validFeatureSplits = getNonConstantFeatures(binAggregates.metadata, featuresForNode)
+    // For each (feature, split), calculate the gain, and select the best (feature, split).
+    val parentImpurityCalc = if (node.stats == null) None else Some(node.stats.impurityCalculator)
```
Note to check: Will node.stats == null for the top level for sure?

smurching replied:
I believe so, the nodes at the top level are created ([RandomForest.scala:178](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala#L178)) with [`LearningNode.emptyNode`](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/tree/Node.scala#L341), which sets `node.stats = null`.

I could change this to check node depth (via node index), but if we're planning on deprecating node indices in the future it might be best not to.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/tree/model/InformationGainStats.scala:
```diff
@@ -112,7 +113,7 @@ private[spark] object ImpurityStats {
    */
   def getInvalidImpurityStats(impurityCalculator: ImpurityCalculator): ImpurityStats = {
-    new ImpurityStats(Double.MinValue, impurityCalculator.calculate(),
+    new ImpurityStats(Double.MinValue, impurity = -1,
```
Q: Why -1 here?

smurching replied:
I changed this to be -1 here since node impurity would eventually get set to -1 anyways when `LearningNodes` with invalid `ImpurityStats` were converted into decision tree leaf nodes (see [`LearningNode.toNode`](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/tree/Node.scala#L279))

jkbradley reviewed (commented):
Done with pass over the parts which refactor elements of RandomForest.scala into utility classes.  Will review more after updates!

smurching committed:
Respond to easy comments

smurching committed:
Merge branch 'master' into pr-splitup

smurching commented on mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala:
```diff
@@ -627,221 +621,37 @@ private[spark] object RandomForest extends Logging {
+  private[impl] def getNonConstantFeatures(
+      metadata: DecisionTreeMetadata,
+      featuresForNode: Option[Array[Int]]): Seq[(Int, Int)] = {
+    Range(0, metadata.numFeaturesPerNode).map { featureIndexIdx =>
```
At some point when refactoring I was hitting errors caused by a stateful operation within a `map` over the output of this method (IIRC the result of the `map` was accessed repeatedly, causing the stateful operation to inadvertently be run multiple times).

However using `withFilter` and `view` now seems to work, I'll change it back :)

smurching committed:
Rename getNonConstantFeatures -> getFeaturesWithSplits (to more accurately reflect what the method actually does). Switch back to view, withFilter in getNonConstantFeatures

smurching commented on mllib/src/main/scala/org/apache/spark/ml/tree/impl/SplitUtils.scala:
```diff
@@ -0,0 +1,215 @@
+      val featureValue = categoriesSortedByCentroid(splitIndex)
+      val leftChildStats =
+        binAggregates.getImpurityCalculator(nodeFeatureOffset, featureValue)
+      val rightChildStats =
```
Exactly, it's the parentCalc minus the left child stats. Since `ImpurityCalculator.subtract()` updates the impurity calculator in place, we call `binAggregates.getParentImpurityCalculator()` to get a copy of the parent impurity calculator, then subtract the left child stats.

smurching committed:
Respond to review comments: compute parent impurity calculator outside the map call in chooseUnorderedCategoricalSplit, orderedSplitHelper

smurching committed:
Merge branch 'master' into pr-splitup

smurching committed:
Remove unneeded newline

SparkQA commented:
**[Test build #83873 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83873/testReport)** for PR 19433 at commit [`0b27c56`](https://github.com/apache/spark/commit/0b27c56d1ea4e1108a62b77e9eca8ae160740756).

smurching committed:
Remove spaces...

SparkQA commented:
**[Test build #83874 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83874/testReport)** for PR 19433 at commit [`d86dd18`](https://github.com/apache/spark/commit/d86dd18e47451c2e4463c68db441f92a898ac765).

SparkQA commented:
**[Test build #83873 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83873/testReport)** for PR 19433 at commit [`0b27c56`](https://github.com/apache/spark/commit/0b27c56d1ea4e1108a62b77e9eca8ae160740756).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83873/
Test FAILed.

SparkQA commented:
**[Test build #83874 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83874/testReport)** for PR 19433 at commit [`d86dd18`](https://github.com/apache/spark/commit/d86dd18e47451c2e4463c68db441f92a898ac765).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83874/
Test FAILed.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/3156/
Test PASSed.

SparkQA commented:
**[Test build #97977 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97977/testReport)** for PR 19433 at commit [`d86dd18`](https://github.com/apache/spark/commit/d86dd18e47451c2e4463c68db441f92a898ac765).

SparkQA commented:
**[Test build #97977 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97977/testReport)** for PR 19433 at commit [`d86dd18`](https://github.com/apache/spark/commit/d86dd18e47451c2e4463c68db441f92a898ac765).
 * This patch **fails to generate documentation**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97977/
Test FAILed.

SparkQA commented:
**[Test build #101549 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101549/testReport)** for PR 19433 at commit [`d86dd18`](https://github.com/apache/spark/commit/d86dd18e47451c2e4463c68db441f92a898ac765).

SparkQA commented:
**[Test build #101549 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101549/testReport)** for PR 19433 at commit [`d86dd18`](https://github.com/apache/spark/commit/d86dd18e47451c2e4463c68db441f92a898ac765).
 * This patch **fails to generate documentation**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101549/
Test FAILed.

SparkQA commented:
**[Test build #101588 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101588/testReport)** for PR 19433 at commit [`d86dd18`](https://github.com/apache/spark/commit/d86dd18e47451c2e4463c68db441f92a898ac765).

holdenk commented:
Is this still a thing you are actively working on?

rstarosta commented:
Thank you for your contribution! We've used this code extensively as a basis for our [@cisco/oraf](https://github.com/cisco/oraf) library, which incorporates local training into the existing decision tree and random forest APIs, and managed to significantly speed-up the training process.

holdenk commented:
That's cool @rstarosta . Does having it in a library meet the needs of folks and we can close this issue?

SparkQA commented:
**[Test build #110569 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110569/testReport)** for PR 19433 at commit [`d86dd18`](https://github.com/apache/spark/commit/d86dd18e47451c2e4463c68db441f92a898ac765).

SparkQA commented:
**[Test build #110569 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110569/testReport)** for PR 19433 at commit [`d86dd18`](https://github.com/apache/spark/commit/d86dd18e47451c2e4463c68db441f92a898ac765).
 * This patch **fails to generate documentation**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110569/
Test FAILed.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/15790/
Test PASSed.
